Failed to build on Max OS X with XCode 5.x llvm toolchain, ld: unknown option: --no-undefined

`./autogen.sh && ./configure && nice -n 20 make -j8` gives:

```
  CCLD     libchewing.la
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/nm: no name list
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/nm: no name list
ld: unknown option: --no-undefined
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libchewing.la] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

And `cmake . && nice -n 20 make -j8` gives:

```
Linking C shared library libchewing.dylib
ld: unknown option: --no-undefined
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libchewing.3.2.0.dylib] Error 1
make[1]: *** [CMakeFiles/chewing_shared.dir/all] Error 2
make: *** [all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,6 +13,8 @@ if(UNIX)
     add_definitions(-DUNDER_POSIX -DPIC)
 endif()
 
+include(CheckCCompilerFlag)
+
 if(${CMAKE_C_COMPILER_ID} STREQUAL "GNU" OR
     ${CMAKE_C_COMPILER_ID} STREQUAL "Clang")
     set(CMAKE_C_FLAGS "-std=gnu99 ${CMAKE_C_FLAGS}")
@@ -25,7 +27,14 @@ if(${CMAKE_C_COMPILER_ID} STREQUAL "GNU" OR
     # Use NO_UNDEFINED=no when running with address sanitizer
     option(NO_UNDEFINED "No undefined symbol in object file" true)
     if(NO_UNDEFINED)
-        set(CMAKE_SHARED_LINKER_FLAGS "-Wl,--no-undefined ${CMAKE_SHARED_LINKER_FLAGS}")
+        set(saved_CMAKE_REQUIRED_FLAGS ${CMAKE_REQUIRED_FLAGS})
+        set(CMAKE_REQUIRED_FLAGS "-Wl,--no-undefined")
+        check_c_compiler_flag("" HAVE_NO_UNDEFINED)
+        set(CMAKE_REQUIRED_FLAGS ${saved_CMAKE_REQUIRED_FLAGS})
+
+        if(HAVE_NO_UNDEFINED)
+            set(CMAKE_SHARED_LINKER_FLAGS "-Wl,--no-undefined ${CMAKE_SHARED_LINKER_FLAGS}")
+        endif()
     endif()
 elseif(CMAKE_C_COMPILER_ID MATCHES "MSVC")
     # XXX: Not sure why we need to use MATCHES instead of STREQUAL here.
@@ -39,7 +48,6 @@ elseif(CMAKE_C_COMPILER_ID MATCHES "MSVC")
     add_definitions(/Dsnprintf=_snprintf /D__func__=__FUNCTION__)
 endif()
 
-include(CheckCCompilerFlag)
 check_c_compiler_flag(-fvisibility=hidden FVISIBILITY_HIDDEN)
 if(${FVISIBILITY_HIDDEN})
     set(CMAKE_C_FLAGS "-fvisibility=hidden ${CMAKE_C_FLAGS}")

--- a/configure.ac
+++ b/configure.ac
@@ -156,6 +156,17 @@ foo () {};
 ])
 AC_MSG_RESULT($has_visibility)
 
+AC_CACHE_CHECK([whether the linker supports --no-undefined],
+               [ac_cv_no_undefined],
+               [
+  saved_LDFLAGS="${LDFLAGS}"
+  LDFLAGS="${LDFLAGS} -Wl,--no-undefined"
+  AC_TRY_LINK([], [], [ac_cv_no_undefined=yes], [ac_cv_no_undefined=no])
+  LDFLAGS="${saved_LDFLAGS}"
+])
+AS_IF([test x$ac_cv_no_undefined = "xyes"], [NO_UNDEFINED_LDFLAGS="-Wl,--no-undefined"])
+AC_SUBST(NO_UNDEFINED_LDFLAGS)
+
 # Platform-dependent
 dnl What kind of system are we using?
 case $host_os in

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -37,7 +37,7 @@ libchewing_la_LDFLAGS = \
 	-version-number $(LIBCHEWING_MAJOR):$(LIBCHEWING_MINOR):$(LIBCHEWING_REVISION) \
 	-rpath $(libdir) \
 	-no-undefined \
-	-Wl,--no-undefined \
+	$(NO_UNDEFINED_LDFLAGS) \
 	-export-symbols-regex "^(chewing)_" \
 	$(NULL)
 

